### PR TITLE
fix Integer floordiv for non-integer denominator; divmod with inf and nan

### LINKED ISSUE #16844: S(2)//S.Half give ZeroDivisionError
In Python, `2//.5 -> 4`

### PR DESCRIPTION
**Pull request by smichr**
#### References to other Issues or PRs

fixes #16844

#### Brief description of what is fixed or changed

Formerly
```python
S(3)//S.Half -> ZeroDivisionError
```
Now
```python
S(3)//S.Half -> 6 matches Python 3//.5 -> 6.0
```

Also correct `divmod` when other is Float to match Python behavior but do not match Python behavior when other is infinite:
```python
>>> divmod(1.,float('-inf'))
(-1.0, -inf)
>>> divmod(1.,float('inf'))
(0.0, 1.0)
>>> divmod(-1.,float('inf'))
(-1.0, inf)
>>> divmod(-1.,float('-inf'))
(0.0, -1.0)
```
In these cases, as with `other == S.NaN`, `(nan, nan)` is being returned.
#### Release Notes

- core
    - numbers: Integer//d corrected for case when d is not an integer
    - divmod(a, b) behavior matches Python's when arguments involve oo or nan

### CODE REVIEW

**Comment by asmeurer** [comment]
Why do we return nans instead of matching the Python behavior for infinities?

**Comment by smichr** [comment]
If `divmod(a, b) -> (w, r)` and means `a == w*b + r` and none of those Python values match then perhaps we should just raise an error at the outset rather than returning NaN's and letting the user find out when they try to use them that they have no meaning.

**Comment by smichr** [comment]
ping @asmeurer - is it ok to just raise an error when oo or nan enter divmod calcuations?

**Comment by smichr** [comment]
For reference, here is how Python behaves:
```python
>>> divmod(int(0),float('nan'))
(nan, nan)
>>> divmod(int(0),float('inf'))
(0.0, 0.0)
>>> divmod(int(0),float('-inf'))
(-0.0, -0.0) -> that "-0.0" tests equal to 0
>>> divmod(int(1),float('inf'))
(0.0, 1.0)
>>> divmod(int(1),float('-inf'))
(-1.0, -inf)
>>> divmod(int(-1),float('inf'))
(-1.0, inf)
>>> divmod(int(-1),float('-inf'))
(0.0, -1.0)
>>> divmod(int(2),float('inf'))
(0.0, 2.0)
>>> divmod(int(2),float('-inf'))
(-1.0, -inf)
>>> divmod(int(-2),float('inf'))
(-1.0, inf)
>>> divmod(int(-2),float('-inf'))
(0.0, -2.0)
```

File: sympy/core/tests/test_numbers.py [hunk 1]
```diff
@@ -176,6 +177,15 @@ def test_divmod():
     assert divmod(S(4), S(-2.1)) == divmod(4, -2.1)
     assert divmod(S(-8), S(-2.5) ) == Tuple(3 , -0.5)
 
+    assert divmod(oo, 1) == (S.NaN, S.NaN)
+    assert divmod(S.NaN, 1) == (S.NaN, S.NaN)
+    assert divmod(1, S.NaN) == (S.NaN, S.NaN)
+    assert divmod(1, oo) == (0, 1)
```
Comment (smichr) [inline sympy/core/tests/test_numbers.py]:
```suggestion
    assert divmod(1, oo) == (0, oo)
```

**Pull request merged**